### DISCUSSION TITLE
Allow migration of custom link types

### DISCUSSION THREAD

**Discussion by pixelmachine** [verbb/hyper#111] [Q&A] [answered]
I've just migrated from Typed Link and most of the fields are good but there are a few errors like:

```
Preparing to migrate field “callToAction” (2f63b7f2-25ba-4846-9c47-296399a68370) content.
    > Unable to convert content #52992 for element #89509
    > Unable to convert content #53000 for element #89780
    > Unable to convert content #53002 for element #89792
    > Unable to convert content #53003 for element #89798
    > Field “callToAction” content migrated.
```

**Comment by pixelmachine** [comment]
All of the failures are custom link types and I see now that the migration only targets the built in types. Perhaps it would be good to have a way to pass additional classes to the migration?

For example just passing the relevant class into the $typeMap works perfectly and imports custom links:

MigrateTypedLinkField.php
MigrateTypedLinkContent.php
```php
public array $typeMap = [
  // ... (default link types) ...
  // My custom link type
  'book' => \modules\booklink\models\BookLinkType::class,
]
```

**Reply by engram-design** [reply, marked as answer]
So yep, there's no handling for custom link types as you'd need to provide a source and destination class to convert them to. But you can of course use an event to register your own handling of these custom types. See [docs](https://verbb.io/craft-plugins/hyper/docs/developers/events#migration-events)

This will modify the `typeMap` property to translate what the old link type was to the new one. Yours would look something like:

```php
use verbb\hyper\events\ModifyMigrationLinkEvent;
use verbb\hyper\migrations\MigrateTypedLinkField;
use verbb\hyper\migrations\MigrateTypedLinkContent;
use yii\base\Event;

Event::on(MigrateTypedLinkField::class, MigrateTypedLinkField::EVENT_MODIFY_LINK_TYPE, function(ModifyMigrationLinkEvent $event) {
    if ($event->oldClass === 'book') {
        $event->newClass = \modules\booklink\models\BookLinkType::class;
    }
});

Event::on(MigrateTypedLinkContent::class, MigrateTypedLinkContent::EVENT_MODIFY_LINK_TYPE, function(ModifyMigrationLinkEvent $event) {
    if ($event->oldClass === 'book') {
        $event->newClass = \modules\booklink\models\BookLinkType::class;
    }
});
```

**Reply by pixelmachine** [reply]
Ah lovely, thanks Josh! It's right in the docs for exactly this and I missed it 🫣
